File: tree-walk-interpreter/go/internal/lox/resolver.go
package lox

type functionType int

const (
	functionTypeNone functionType = iota
	functionTypeFunction
)

type Resolver struct {
	interpreter         *Interpreter
	scopes              *stack
	currentFunctionType functionType
}

func NewResolver(interpreter *Interpreter) *Resolver {
	return &Resolver{
		interpreter:         interpreter,
		scopes:              newStack(),
		currentFunctionType: functionTypeNone,
	}
}

func (r *Resolver) Resolve(statements []Stmt) {
	r.resolveStatements(statements)
}

func (r *Resolver) VisitAssignExpr(expr *Assign) interface{} {
	r.resolveExpression(expr.Value)
	r.resolveLocal(expr, expr.Name)
	return nil
}

func (r *Resolver) VisitBinaryExpr(expr *Binary) interface{} {
	r.resolveExpression(expr.Left)
	r.resolveExpression(expr.Right)
	return nil
}

func (r *Resolver) VisitCallExpr(expr *Call) interface{} {
	r.resolveExpression(expr.Callee)

	for _, argument := range expr.Arguments {
		r.resolveExpression(argument)
	}

	return nil
}

func (r *Resolver) VisitGetExpr(expr *Get) interface{} {
	r.resolveExpression(expr.Object)
	return nil
}

func (r *Resolver) VisitGroupingExpr(expr *Grouping) interface{} {
	r.resolveExpression(expr.Expression)
	return nil
}

func (r *Resolver) VisitLiteralExpr(expr *Literal) interface{} {
	return nil
}

func (r *Resolver) VisitLogicalExpr(expr *Logical) interface{} {
	r.resolveExpression(expr.Left)
	r.resolveExpression(expr.Right)
	return nil
}

func (r *Resolver) VisitSetExpr(expr *Set) interface{} {
	r.resolveExpression(expr.Value)
	r.resolveExpression(expr.Object)
	return nil
}

func (r *Resolver) VisitSuperExpr(expr *Super) interface{} {
	return nil
}

func (r *Resolver) VisitThisExpr(expr *This) interface{} {
	return nil
}

func (r *Resolver) VisitUnaryExpr(expr *Unary) interface{} {
	r.resolveExpression(expr.Right)
	return nil
}

func (r *Resolver) VisitVariableExpr(expr *Variable) interface{} {
	if !r.scopes.empty() {
		if declared, ok := r.scopes.peek()[expr.Name.Lexeme]; ok && !declared {
			newParseError(expr.Name, "Can't read local variable in its own initializer.")
		}
	}

	r.resolveLocal(expr, expr.Name)
	return nil
}

func (r *Resolver) VisitBlockStmt(stmt *Block) interface{} {
	r.beginScope()
	r.resolveStatements(stmt.Statements)
	r.endScope()
	return nil
}

func (r *Resolver) VisitClassStmt(stmt *Class) interface{} {
	r.declare(stmt.Name)
	r.define(stmt.Name)
	return nil
}

func (r *Resolver) VisitExpressionStmt(stmt *Expression) interface{} {
	r.resolveExpression(stmt.Expression)
	return nil
}

func (r *Resolver) VisitFunctionStmt(stmt *Function) interface{} {
	r.declare(stmt.Name)
	r.define(stmt.Name)

	r.resolveFunction(stmt, functionTypeFunction)
	return nil
}

func (r *Resolver) VisitIfStmt(stmt *If) interface{} {
	r.resolveExpression(stmt.Condition)
	r.resolveStatement(stmt.ThenBranch)
	if stmt.ElseBranch != nil {
		r.resolveStatement(stmt.ElseBranch)
	}
	return nil
}

func (r *Resolver) VisitPrintStmt(stmt *Print) interface{} {
	r.resolveExpression(stmt.Expression)
	return nil
}

func (r *Resolver) VisitReturnStmt(stmt *Return) interface{} {
	if r.currentFunctionType == functionTypeNone {
		newParseError(stmt.Keyword, "Can't return from top-level code.")
	}

	if stmt.Value != nil {
		r.resolveExpression(stmt.Value)
	}

	return nil
}

func (r *Resolver) VisitVarStmt(stmt *Var) interface{} {
	r.declare(stmt.Name)
	if stmt.Initializer != nil {
		r.resolveExpression(stmt.Initializer)
	}
	r.define(stmt.Name)
	return nil
}

func (r *Resolver) VisitWhileStmt(stmt *While) interface{} {
	r.resolveExpression(stmt.Condition)
	r.resolveStatement(stmt.Body)
	return nil
}

func (r *Resolver) resolveExpression(expr Expr) {
	expr.Accept(r)
}

func (r *Resolver) resolveStatement(statement Stmt) {
	statement.Accept(r)
}

func (r *Resolver) resolveStatements(statements []Stmt) {
	for _, statement := range statements {
		r.resolveStatement(statement)
	}
}

func (r *Resolver) resolveLocal(expr Expr, name *Token) {
	l := len(r.scopes.values)
	for i := l - 1; i >= 0; i-- {
		if _, ok := r.scopes.values[i][name.Lexeme]; ok {
			r.interpreter.resolve(expr, l-1-i)
			return
		}
	}
}

func (r *Resolver) resolveFunction(function *Function, functionType functionType) {
	enclosingFunctionType := r.currentFunctionType
	r.currentFunctionType = functionType
	r.beginScope()
	for _, param := range function.Parameters {
		r.declare(param)
		r.define(param)
	}
	r.resolveStatements(function.Body)
	r.endScope()
	r.currentFunctionType = enclosingFunctionType
}

func (r *Resolver) beginScope() {
	r.scopes.push(newScope())
}

func (r *Resolver) endScope() {
	r.scopes.pop()
}

func (r *Resolver) declare(name *Token) {
	if r.scopes.empty() {
		return
	}

	scope := r.scopes.peek()
	if _, ok := scope[name.Lexeme]; ok {
		newParseError(name, "Already a variable with this name in this scope.")
	}

	scope[name.Lexeme] = false
}

func (r *Resolver) define(name *Token) {
	if r.scopes.empty() {
		return
	}
	r.scopes.peek()[name.Lexeme] = true
}

File: tree-walk/cs/Lox/Interpreter/Resolver.cs
﻿using System;
using Lox.Parser;
using Lox.Scanner;

namespace Lox.Interpreter;

class Resolver : IExprVisitor<Void?>, IStmtVisitor<Void?>
{
    private enum FunctionType
    {
        None,
        Function,
        Initializer,
        Method,
    }

    private enum ClassType
    {
        None,
        Class,
        Subclass,
    }

	private readonly Interpreter interpreter;
    private readonly Stack<Dictionary<string, bool>> scopes = new Stack<Dictionary<string, bool>>();
    private FunctionType currentFunctionType = FunctionType.None;
    private ClassType currentClassType = ClassType.None;

	public Resolver(Interpreter interpreter)
	{
		this.interpreter = interpreter;
	}

    public void Resolve(List<Stmt> statements)
    {
        foreach (var statement in statements)
        {
            Resolve(statement);
        }
    }

    public Void? VisitAssignExpr(Assign expr)
    {
        Resolve(expr.Value);
        ResolveLocal(expr, expr.Name);
        return null;
    }

    public Void? VisitBinaryExpr(Binary expr)
    {
        Resolve(expr.Left);
        Resolve(expr.Right);
        return null;
    }

    public Void? VisitCallExpr(Call expr)
    {
        Resolve(expr.Callee);

        foreach (var argument in expr.Arguments)
        {
            Resolve(argument);
        }

        return null;
    }

    public Void? VisitGetExpr(Get expr)
    {
        Resolve(expr.Object);
        return null;
    }

    public Void? VisitGroupingExpr(Grouping expr)
    {
        Resolve(expr.Expression);
        return null;
    }

    public Void? VisitLiteralExpr(Literal expr)
    {
        return null;
    }

    public Void? VisitLogicalExpr(Logical expr)
    {
        Resolve(expr.Left);
        Resolve(expr.Right);
        return null;
    }

    public Void? VisitSetExpr(Set expr)
    {
        Resolve(expr.Value);
        Resolve(expr.Object);
        return null;
    }

    public Void? VisitSuperExpr(Super expr)
    {
        if (currentClassType == ClassType.None)
        {
            Program.Error(expr.Keyword, "Can't use 'super' outside of a class.");
        }
        else if (currentClassType != ClassType.Subclass)
        {
            Program.Error(expr.Keyword, "Can't use 'super' in a class with no superclass.");
        }

        ResolveLocal(expr, expr.Keyword);
        return null;
    }

    public Void? VisitThisExpr(This expr)
    {
        if (currentClassType == ClassType.None)
        {
            Program.Error(expr.Keyword, "Can't use 'this' outside of a class.");
            return null;
        }

        ResolveLocal(expr, expr.Keyword);
        return null;
    }

    public Void? VisitUnaryExpr(Unary expr)
    {
        Resolve(expr.Right);
        return null;
    }

    public Void? VisitVariableExpr(Variable expr)
    {
        if (scopes.Count > 0 &&
            scopes.Peek().TryGetValue(expr.Name.Lexeme, out var defined))
        {
            if (defined == false)
            {
                Program.Error(expr.Name, "Can't read local variable in its own initializer.");
            }
        }

        ResolveLocal(expr, expr.Name);
        return null;
    }

    public Void? VisitBlockStmt(Block stmt)
    {
        BeginScope();
        Resolve(stmt.Statements);
        EndScope();
        return null;
    }

    public Void? VisitClassStmt(Class stmt)
    {
        var enclosingClassType = currentClassType;
        currentClassType = ClassType.Class;

        Declare(stmt.Name);
        Define(stmt.Name);

        if (stmt.Superclass != null &&
            stmt.Name.Lexeme == stmt.Superclass.Name.Lexeme)
        {
            Program.Error(stmt.Superclass.Name, "A class can't inherit from itself.");
        }

        if (stmt.Superclass != null)
        {
            currentClassType = ClassType.Subclass;
            Resolve(stmt.Superclass);
        }

        if (stmt.Superclass != null)
        {
            BeginScope();
            scopes.Peek()["super"] = true;
        }

        BeginScope();
        scopes.Peek()["this"] = true;

        foreach (var method in stmt.Methods)
        {
            var declaration = FunctionType.Method;
            if (method.Name.Lexeme == "init")
            {
                declaration = FunctionType.Initializer;
            }
            ResolveFunction(method, declaration);
        }

        EndScope();

        if (stmt.Superclass != null)
        {
            EndScope();
        }

        currentClassType = enclosingClassType;
        return null;
    }

    public Void? VisitExpressionStmt(ExpressionStmt stmt)
    {
        Resolve(stmt.Expression);
        return null;
    }

    public Void? VisitFunctionStmt(Function stmt)
    {
        Declare(stmt.Name);
        Define(stmt.Name);

        ResolveFunction(stmt, FunctionType.Function);
        return null;
    }

    public Void? VisitIfStmt(If stmt)
    {
        Resolve(stmt.Condition);
        Resolve(stmt.ThenBranch);
        if (stmt.ElseBranch != null)
        {
            Resolve(stmt.ElseBranch);
        }
        return null;
    }

    public Void? VisitPrintStmt(Print stmt)
    {
        Resolve(stmt.Expression);
        return null;
    }

    public Void? VisitReturnStmt(Return stmt)
    {
        if (currentFunctionType == FunctionType.None)
        {
            Program.Error(stmt.Keyword, "Can't return from top-level code.");
        }

        if (stmt.Value != null)
        {
            if (currentFunctionType == FunctionType.Initializer)
            {
                Program.Error(stmt.Keyword, "Can't return a value from an initializer.");
            }

            Resolve(stmt.Value);
        }

        return null;
    }

    public Void? VisitVarStmt(Var stmt)
    {
        Declare(stmt.Name);
        if (stmt.Initializer != null)
        {
            Resolve(stmt.Initializer);
        }
        Define(stmt.Name);
        return null;
    }

    public Void? VisitWhileStmt(While stmt)
    {
        Resolve(stmt.Condition);
        Resolve(stmt.Body);
        return null;
    }

    private void Resolve(Stmt stmt)
    {
        stmt.Accept(this);
    }

    private void Resolve(Expr expr)
    {
        expr.Accept(this);
    }

    private void BeginScope()
    {
        scopes.Push(new Dictionary<string, bool>());
    }

    private void EndScope()
    {
        scopes.Pop();
    }

    private void Declare(Token name)
    {
        if (scopes.Count == 0)
        {
            return;
        }

        var scope = scopes.Peek();
        if (scope.ContainsKey(name.Lexeme))
        {
            Program.Error(name, "Already a variable with this name in this scope.");
        }

        scope[name.Lexeme] = false;
    }

    private void Define(Token name)
    {
        if (scopes.Count == 0)
        {
            return;
        }
        scopes.Peek()[name.Lexeme] = true;
    }

    private void ResolveLocal(Expr expr, Token name)
    {
        var distance = 0;
        foreach (var scope in scopes)
        {
            if (scope.ContainsKey(name.Lexeme))
            {
                interpreter.Resolve(expr, distance);
                return;
            }

            distance++;
        }
    }

    private void ResolveFunction(Function function, FunctionType functionType)
    {
        var enclosingFunctionType = currentFunctionType;
        currentFunctionType = functionType;

        BeginScope();
        foreach (var param in function.Parameters)
        {
            Declare(param);
            Define(param);
        }
        Resolve(function.Body);
        EndScope();

        currentFunctionType = enclosingFunctionType;
    }
}

